### DISCUSSION TITLE
How to link OS and BSP implementation files into the OSAL using the CMakelists

### DISCUSSION THREAD

**Discussion by spaceusername** [nasa/cFS#761] [Q&A] [answered]
Hey all, 

I'm currently trying to integrate a new target bsp and RTOS into the cFS build. I see in the os-impl files for rtems and vxworks that angle bracket includes are used for OS source files etc. I have not been able to find where the included files are set as SYSTEM includes in the CMakelists. Is  include_directories(SYSTEM ......) used or is another method being used that I am not aware of?

My chief concern is that I would like to integrate the new BSP, OS, Filesystem, and Network Layer appropriately, but I have not been able to find many resources that detail that information when it comes to linking in the actual source files of each. So far I have been checking through the discussion forums as well as other existent forks of cFS. 

Any information yall can provide is appreciated.

**Comment by skliper** [comment]
Is this what you are looking for?

https://github.com/nasa/PSP/blob/70be39db9259061d3c9aaf1bc3eb8921e8794828/cmake/Modules/Platform/RTEMS.cmake#L87-L89

Typically it's either in the toolchain or one of these platform files if this isn't the exact setting you are interested in.

**Reply by spaceusername** [reply]
Yes, I think this line specifically:
set(CMAKE_SYSTEM_INCLUDE_PATH
    "${RTEMS_TARGET_PATH}/${RTEMS_BSP}/lib/include"
    "${RTEMS_TOOLS_PATH}/include")
    
I'm pretty much looking for where I link in the implementation source for the RTOS. 

Is the preferred method for linking in the network and filesystem source files to also add them to the system include path?

**Reply by skliper** [reply, marked as answer]
Ohhhh, for the current project I'm working linking source for RTEMS depends on if you want to make a stand-alone executable or a loadable module.  To support osal unit testing with stand-alone executables it needs to be linked in from the OSAL BSP:
https://github.com/nasa/osal/blob/372ea659885d9b306bfdb9b879e0d56c74f96709/src/bsp/generic-rtems/CMakeLists.txt#L26-L31

If just loading it, then VxWorks and RTEMS have the implementation source linked in wherever you are loading from.  Related PSP logic:
https://github.com/nasa/PSP/blob/70be39db9259061d3c9aaf1bc3eb8921e8794828/cmake/Modules/Platform/RTEMS.cmake#L45-L72

**Reply by spaceusername** [reply]
Yeah, the setup that I am currently working on is  freeRTOS which does not, by default, have a filesystem or network layer. I'm using appropriate libraries to extend it. The end processor target will probably need a standalone single file for execution. I don't think I will be able to do a dynamic load given the constraints of the operating system. 

I have already taken a look at the METEC fork that was made, but was unclear primarily on the link step.

I'm attempting to determine the CMakelists I need to edit to get the source files for those libraries and the OS source files into the include structure.

**Reply by skliper** [reply]
If you want any of the osal functional tests that rely on the file system or network to work as stand-alone executables I recommend linking that stuff in from the OSAL BSP.

**Reply by spaceusername** [reply]
Ok, I'll try to go that route then. I do have a question about the rtemscpu link target. I see in the i686-rtems5.cmake that the files are linked in with this line:

set(LINK_LIBRARIES              "-lrtemsdefaultconfig -lrtemsbsp -lrtemscpu")

Is that for the bsp and cpu setup as installed libraries in the build system, or are they passed in from somewhere else? I'm hoping to link in the filesystem and network layer such that the build calls into their cmakelists for make generation and build. Would I need to consider a different approach for that?

I also have a clarifying question on the PSP. I'm of the understanding that the PSP primarily exists to glue together the cFE and BSP functions. Further, from what you're saying, if I link in the RTOS, filesystem, and network source in the OSAL BSP I should not need to worry about linking it in in the PSP. Is that correct?

**Reply by spaceusername** [reply]
I've marked one of your earlier replies as the answer, though I think the comments as a whole compose the complete answer. After a little bit of reading the cmake documentation I have answered the rest of my questions related to linking in what I need to. I could still use a little clarification on linking in the file system and network layer linking, but I will post a separate question for that.
